msfextract

```diff
@@ -0,0 +1,240 @@
+  banner << "Example: #{$0} -a metasploit.apk -o /root/configbytes.txt -j /root/fernflower.jar"
+  opt.banner = banner
+  opt.separator('')
+  opt.separator('Options:')
+
+  opt.on('-a', '--apk             <path>', String, 'Specify apk to extract configbytes') do |a|
+    opts[:apk] = a
+  end
+
+  opt.on('-o', '--out             <path>', String, 'Save configbytes to a file') do |o|
+    opts[:out] = o
+  end
+
+  opt.on('-j', '--jar             <path>', String, 'Specify fernflower path') do |j|
+    opts[:jar] = j
+  end
+
+  opt.on('-v', '--verbose', 'Displays verbose output') do
+    opts[:verbose] = true
+  end
+
+  opt.on('-k', '--keep', 'Keep working directory') do
+    opts[:keep] = true
+  end
+
+
+  opt.on_tail('-h', '--help', 'Show this message') do
+    raise HelpError, "#{opt}"
+  end
+
+  begin
+    opt.parse!(args)
+  rescue OptionParser::InvalidOption => e
+    raise UsageError, "Invalid option\n#{opt}"
+  rescue OptionParser::MissingArgument => e
+    raise UsageError, "Missing required argument for option\n#{opt}"
+  end
+
+  if opts.empty?
+    raise UsageError, "No options\n#{opt}"
+  end
+
+  if opts[:apk].nil?
+    raise UsageError, "Missing required argument apk file\n#{opt}"
+    exit(1)
+  end
+
+  if opts[:jar].nil?
+    raise UsageError, "Missing required argument fernflower path\n#{opt}"
+    exit(1)
+  end
+
+  opts
+end
+
+begin
+  options = parse_args(ARGV)
+rescue HelpError => e
+  $stderr.puts e.message
+  exit(1)
+rescue MsfExtractError => e
+  $stderr.puts "Error: #{e.message}"
+  exit(1)
+end
+
+begin
+  check_tools()
+
+  temp_dir = Dir.mktmpdir
+  temp_dir << "/"
+
+  if options[:out].nil?
+    $stderr.puts "[-] No output option selected, working in #{temp_dir}"    
+    output = "#{temp_dir}configbytes.txt"
+    options[:keep] = true
+  else
+    output = options[:out]
+  end
+
+  apk = options[:apk]
+  zip_file = "#{temp_dir}temp.zip"
+  temp_apk = "#{temp_dir}temp.apk"
+  FileUtils.cp(apk, temp_apk)
+
+  $stderr.puts "[+] Renaming apk file to zip file"
+  File.rename(temp_apk, zip_file)
+
+  $stderr.puts "[+] Using unzip on zip file for dex file"
+  unzip_cmd = run_cmd("unzip #{temp_dir}temp.zip -d #{temp_dir}")
+  if options[:verbose] then $stderr.puts "#{unzip_cmd}" else unzip_cmd end
+
+  $stderr.puts "[+] Using d2j-dex2jar on dex file to create jar file"
+  dex_file = "#{temp_dir}classes.dex"
+  d2j_cmd =  run_cmd("d2j-dex2jar -o #{temp_dir}classes.jar #{dex_file}")
+  if options[:verbose] then $stderr.puts "#{d2j_cmd}" else d2j_cmd end
+
+  $stderr.puts "[+] Using unzip on jar file for class files"
+  classes_dir = "#{temp_dir}classes/"
+  FileUtils.mkdir_p classes_dir
+  unzip_jar_cmd = run_cmd("unzip #{temp_dir}classes.jar -d #{classes_dir}")
+  if options[:verbose] then $stderr.puts "#{unzip_jar_cmd}" else unzip_jar_cmd end
+  
+  $stderr.puts "[+] Using apktool to read AndroidManifest"
+  apktool_cmd = run_cmd("apktool d #{apk} -o #{temp_dir}decompile/")
```

Comment:
This will fail if the api filename has spaces in it FYI. You should really pass the arguments to all of these command invocations as an array to avoid that kind of issue.